# **DATA CLEANING CHALLENGE**

Authors: Ryan Musasia 

## Importing Libraries

Libraries to enhance data manipulation

These comprises of;
   > **Pandas** 

   Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
   built on top of the Python programming language.

   > **Numpy** 

   NumPy is an open source project aiming to enable numerical computing with Python.

In [1]:
import pandas as pd 
import numpy as np 

Libraries for visualization

These comprise of;

   > **Matplotlib**

   Matplotlib is a comprehensive library for creating static, animated, and interactive visualizations in Python.

   > **Seaborn**

   Seaborn is a Python data visualization library based on matplotlib. It provides a high-level interface for drawing attractive and informative statistical graphics.

In [2]:
import matplotlib.pyplot as plt 
import seaborn as sns 

## Loading data 

In [3]:
# Loading the data using pandas into df {Data frame}

# ID sets the index to be the player id as there is no need to have more than one unique identifiers on the first dataset.

df = pd.read_csv(r"C:\Users\user\OneDrive\Desktop\Projects\challenge\data-challenge\data\fifa21 raw data v2.csv")

C:\Users\user\AppData\Local\Temp\ipykernel_29396\3637115478.py:5: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r"C:\Users\user\OneDrive\Desktop\Projects\challenge\data-challenge\data\fifa21 raw data v2.csv")


# 2. DATA UNDERSTANDING

## Overview

We'll be working with a messy FIFA dataset  

The data has already been loaded in the above code cell.

In [4]:
#preview the first 5 rows of this dataset
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

In [5]:
#preview the last 5 rows of the data 
df.tail()

ID          Name        LongName  \
18974  247223        Xia Ao          Ao Xia   
18975  258760      B. Hough       Ben Hough   
18976  252757   R. McKinley  Ronan McKinley   
18977  243790  Wang Zhen'ao    Zhen'ao Wang   
18978  252520     Zhou Xiao       Xiao Zhou   

                                               photoUrl  \
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                      Club  ...     A/W     D/W   IR PAC SHO  \
18974    47   55        \n\n\n\nWuhan Zall  ...  Medium  Medium  1 ★  64  28   
18975    47   67   \n\n\n\nOldham Athletic  ...  Medium  Medium  1 ★  64  40   
18976    47   65        \n\n\n\nDerry City  ...  Medium  Medium  1 ★  63  39   
18977    47   57  \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  58  49   
18978    47   57  \n\n\n\nDalian YiFang FC  ...  Medium  Medium  1 ★  62  22   

       PAS DRI DEF PHY Hits  
18974   26  38  48  51  NaN  
18975   48  49  35  45  NaN  
18976   44  46  40  53  NaN  
18977   41  49  30  44  NaN  
18978   39  42  45  55  NaN  

[5 rows x 77 columns]

In [6]:
#create a function that will be able to perform data understanding 
class DataUnderstanding(object):
    '''a class that does basic data understanding'''
    def __init__(self, df):
        self.shape = df.shape
        self.info = df.info
        self.types = df.dtypes 
        self.describe = df.describe
        self.duplicates = df.duplicated().sum()
        self.missing = df.isna().sum()

In [7]:
#instantiating the class
data_und = DataUnderstanding(df)

In [8]:
print(f"This dataframe has the following shape:{data_und.shape}")

This dataframe has the following shape:(18979, 77)


In [9]:
data_und.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

In [10]:
data_und.describe()

ID           Age          ↓OVA           POT           BOV  \
count   18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean   226403.384794     25.194109     65.718636     71.136414     66.751726   
std     27141.054157      4.710520      6.968999      6.114635      6.747193   
min        41.000000     16.000000     47.000000     47.000000     48.000000   
25%    210135.000000     21.000000     61.000000     67.000000     62.000000   
50%    232418.000000     25.000000     66.000000     71.000000     67.000000   
75%    246922.500000     29.000000     70.000000     75.000000     71.000000   
max    259216.000000     53.000000     93.000000     95.000000     93.000000   

          Attacking      Crossing     Finishing  Heading Accuracy  \
count  18979.000000  18979.000000  18979.000000      18979.000000   
mean     248.938142     49.688392     45.842405         51.942726   
std       74.299428     18.131153     19.567081         17.294409   
min       42.000000      6.000000      3.000000          5.000000   
25%      222.000000     38.000000     30.000000         44.000000   
50%      263.000000     54.000000     49.000000         55.000000   
75%      297.000000     63.000000     62.000000         64.000000   
max      437.000000     94.000000     95.000000         93.000000   

       Short Passing  ...  GK Positioning   GK Reflexes   Total Stats  \
count   18979.000000  ...    18979.000000  18979.000000  18979.000000   
mean       58.768112  ...       16.217187     16.519627   1595.286949   
std        14.519106  ...       17.002239     17.854079    269.874789   
min         7.000000  ...        2.000000      2.000000    747.000000   
25%        54.000000  ...        8.000000      8.000000   1452.000000   
50%        62.000000  ...       11.000000     11.000000   1627.000000   
75%        68.000000  ...       14.000000     14.000000   1781.000000   
max        94.000000  ...       91.000000     90.000000   2316.000000   

         Base Stats           PAC           SHO           PAS           DRI  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean     355.702197     67.453975     53.457031     57.681016     62.875020   
std       40.761117     10.677859     13.827425     10.081857      9.927415   
min      232.000000     25.000000     16.000000     25.000000     25.000000   
25%      327.000000     61.000000     44.000000     51.000000     57.000000   
50%      356.000000     68.000000     56.000000     58.000000     64.000000   
75%      384.000000     75.000000     64.000000     64.000000     69.000000   
max      498.000000     96.000000     93.000000     93.000000     95.000000   

                DEF           PHY  
count  18979.000000  18979.000000  
mean      49.866221     64.368934  
std       16.443213      9.601883  
min       12.000000     28.000000  
25%       35.000000     58.000000  
50%       53.000000     65.000000  
75%       63.000000     71.000000  
max       91.000000     91.000000  

[8 rows x 54 columns]

In [11]:
print(f'The number of duplicated values in the datarame: {data_und.duplicates}')

The number of duplicated values in the datarame: 0


In [12]:
print(f'The number of missing values in the dataframe: {data_und.missing}')

The number of missing values in the dataframe: ID              0
Name            0
LongName        0
photoUrl        0
playerUrl       0
             ... 
PAS             0
DRI             0
DEF             0
PHY             0
Hits         2595
Length: 77, dtype: int64


The Hits column contains 2595 missing values 

The ```Hits``` column tracks the popularity of search terms over time, or social media platforms such as Tsuch as the official FIFA website or the FIFA Ultimate Team (FUT) online marketplace, which allows users to buy and sell player cards for the game.

__Summary__

* Our data has 18,979 rows and 77 columns
* The data has 54 columns of dtype ```int64``` and 23 columns of dtype ```object```.
* There are missing values in the ```Hits``` columns of the data.
* There are no duplicates in the data.
* The following are not relevant columns in our data;
    * photoUrl 
    * playerUrl
    * Loan Date End
    * Release Clause


Since our data has missing values, we therefore ought to remove them in the data cleaning process

## DATA CLEANING

In [13]:
#dropping irrelevant columns 
df = df.drop(columns=['photoUrl', 'playerUrl', 'Loan Date End', 'Release Clause'])
df.head()

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   ↓OVA  POT                         Club     Contract   Positions  ...  \
0    93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  ...   
1    92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  ...   
2    91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  ...   
3    91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  ...   
4    91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  ...   

      A/W     D/W   IR  PAC SHO PAS DRI DEF  PHY  Hits  
0  Medium     Low  5 ★   85  92  91  95  38   65   771  
1    High     Low  5 ★   89  93  81  89  35   77   562  
2  Medium  Medium  3 ★   87  92  78  90  52   90   150  
3    High    High  4 ★   76  86  93  88  64   78   207  
4    High  Medium  5 ★   91  85  86  94  36   59   595  

[5 rows x 73 columns]

In [14]:
#removing the unnecessary newline characters from all column names that have them
df.columns = df.columns.str.replace('\n', '')

# print updated column names
print(df.columns)

Index(['ID', 'Name', 'LongName', 'Nationality', 'Age', '↓OVA', 'POT', 'Club',
       'Contract', 'Positions', 'Height', 'Weight', 'Preferred Foot', 'BOV',
       'Best Position', 'Joined', 'Value', 'Wage', 'Attacking', 'Crossing',
       'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill',
       'Dribbling', 'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control',
       'Movement', 'Acceleration', 'Sprint Speed', 'Agility', 'Reactions',
       'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength',
       'Long Shots', 'Mentality', 'Aggression', 'Interceptions', 'Positioning',
       'Vision', 'Penalties', 'Composure', 'Defending', 'Marking',
       'Standing Tackle', 'Sliding Tackle', 'Goalkeeping', 'GK Diving',
       'GK Handling', 'GK Kicking', 'GK Positioning', 'GK Reflexes',
       'Total Stats', 'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC',
       'SHO', 'PAS', 'DRI', 'DEF', 'PHY', 'Hits'],
      dtype='object')


In [15]:
#identifying the columns with missing values 
def missing_cols(df):
    '''prints out columns with its amount of missing values'''
    total = 0
    for col in df.columns:
        missing_vals = df[col].isnull().sum()
        total += missing_vals
        if missing_vals != 0:
            print(f"{col} => {df[col].isnull().sum()}")
    
    if total == 0:
        print("no missing values left")
            
missing_cols(df)

Hits => 2595


Only the ```Hits``` column has missing values which account for only 13% of the total data, I chose to remove the rows containing the missing values.

In [16]:
df = df.dropna()

In [17]:
data_und = DataUnderstanding(df)
data_und.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16384 entries, 0 to 16636
Data columns (total 73 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                16384 non-null  int64 
 1   Name              16384 non-null  object
 2   LongName          16384 non-null  object
 3   Nationality       16384 non-null  object
 4   Age               16384 non-null  int64 
 5   ↓OVA              16384 non-null  int64 
 6   POT               16384 non-null  int64 
 7   Club              16384 non-null  object
 8   Contract          16384 non-null  object
 9   Positions         16384 non-null  object
 10  Height            16384 non-null  object
 11  Weight            16384 non-null  object
 12  Preferred Foot    16384 non-null  object
 13  BOV               16384 non-null  int64 
 14  Best Position     16384 non-null  object
 15  Joined            16384 non-null  object
 16  Value             16384 non-null  object
 17  Wage        

In [18]:
#convert the height and weight columns to numerical forms
